### PR TITLE
Allow cross-platform CAPI-compatible remote key exchange

### PR DESCRIPTION
**Pull request by silijon**
# PR Summary

Creates a dotnet core native implementation of the CryptoUtils key exchange enabling cross-platform remote sessions to successfully serialize SecureString.

## PR Context

Currently, after importing a remote session between a Windows and a Linux host (in either direction), any imported cmdlets parameterized with SecureString are broken-by-design. There is a lazily triggered operation that happens during a SecureString serialization that uses the windows native CAPI crypto functions (e.g. CryptGenKey) to do the following:

1. Generate a public/private keypair.
2. Export/import the public key.
3. Generate a symmetric session key.
4. Encrypt and exchange the session key.
5. Encrypt the SecureString for transit.

The implementation of this protocol is conditioned on the UNIX compile flag and, for non-Windows builds, it simply throws a PSCryptoException. The result is that any and all cmdlets that use a SecureString parameter are broken for cross-platform importation/invocation.

This PR fixes this problem by doing the following:

1. Extracts an interface from _PSRSACryptoServiceProvider_.

2. Re-names the current _PSRSACryptoServiceProvider_ to _Win32PSRSACryptoServiceProvider_.

3. Implements _CorePSRSACryptoServiceProvider_ with the same external semantics as _Win32PSRSACryptoServiceProvider_ but uses the core-compatible AES and RSA classes to do key generation, importation, exportation, encrypt, and decrypt.

4. Uses a small set of conversions to enable _CorePSRSACryptoServiceProvider_ to return/digest the key blob formats to/from the documented CAPI formats so as to retain compatibility with native windows systems.

5. Conditions the use of _CorePSRSACryptoServiceProvider_ on the UNIX compile flag.

*Notes* 

- I looked for an issue specifically for this problem and, while there are a number of closed issues relating to more general SecureString problems (having to do with the DPAPI and other things) I didn't see one directly related to this. Would be happy to tag an issue in here if I just missed it.

- This PR fixes the same issue raised with the PS team in discussion with the SkyKick team.  Specifically @PaulHigin (via email) and @TylerLeonhardt (on his livestream).

## PR Checklist

- [X] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [X] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [X] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [X] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [X] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [X] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

**Comment by TylerLeonhardt** [comment]
This will likely fix the local debugging experience for Azure Functions on macOS and Linux which catastrophically fails when you store a SecureString to a variable which triggers the key exchange.

Im excited at the impact this change could have!

**Comment by PaulHigin** [comment]
SecureString is not implemented with encryption on non-Windows platforms.  So encryption with the symmetric key over the wire seems unneeded.  It may be better to remove it entirely.  Adding committee review tag.

**Comment by silijon** [comment]
Yeah, session encryption during serialization is entirely redundant if the transport is encrypted. Here I was trying to make the smallest change possible while maintaining compatibility with older version remote shells.

Another thing for consideration: I think the core AES/RSA classes fall back on the windows native libraries in their implementation?  If so, you may consider just dropping Win32PSRSACryptoServiceProvider altogether.

**Comment by PaulHigin** [comment]
Yeah, I am beginning to see your point.  This may be the best way moving forward and keeping backward compatibility.  There is still the deadlock issue when server initiates key exchange, but I think that can be fixed separately.

I would like to get rid of `Win32PSRSACryptoServiceProvider`.  Even if dotNet core implementation does not fall back, is there a reason to keep it?  Isn't the core version compatible ... it would need to be for full x-plat compatibility to older versions of PowerShell.

**Comment by silijon** [comment]
Personally, I think getting rid of `Win32PSRSACryptoServiceProvider` is the right move but I wasn't sure if you guys had a reason to leave the native windows crypto stuff in.  So I tried to do as light a touch as possible.

I'd be happy to refactor it out.  It'll allow a lot of code and compiler conditions to be removed from `CryptoUtils` and simplify the whole thing.  I think it's a good trade-off to ensure backward compat but clean some of this stuff up.

**Comment by silijon** [comment]
I updated this PR to completely remove the native windows dll references and change the entire key exchange over to using dotnet core compatible aes and rsa classes.  I also consolidated the CAPI conversion code into CryptoUtils.cs and used the existing pattern of byte constant naming to be more explicit about the structure of the keyblob format conversions.  Let me know what you think.

**Comment by PaulHigin** [comment]
Thanks @silijon, I'll take a look this week.

File: src/System.Management.Automation/utils/CryptoUtils.cs
Comment (PaulHigin) [inline src/System.Management.Automation/utils/CryptoUtils.cs]: Localizable messages need to go into `SecuritySupportStrings.resx`.

File: src/System.Management.Automation/utils/CryptoUtils.cs
Comment (PaulHigin) [inline src/System.Management.Automation/utils/CryptoUtils.cs]: Localizable messages need to go into `SecuritySupportStrings.resx`.

File: src/System.Management.Automation/utils/CryptoUtils.cs
Comment (PaulHigin) [inline src/System.Management.Automation/utils/CryptoUtils.cs]: Should there be a check or assert here for (n <= 4)?  Otherwise exponent will run into modulus...

File: src/System.Management.Automation/utils/CryptoUtils.cs
Comment (PaulHigin) [inline src/System.Management.Automation/utils/CryptoUtils.cs]: Don't need index and length here since we are reversing the entire array. (`Array.Reverse(part);`).

File: src/System.Management.Automation/utils/CryptoUtils.cs
Comment (PaulHigin) [inline src/System.Management.Automation/utils/CryptoUtils.cs]: I think this would be more readable if `part` name was `modulus`.  Also `len` is not needed since it is the same as `keyLength` above.
```
Buffer.BlockCopy(modulus, 0, blob, pos, keyLength);
```

File: src/System.Management.Automation/utils/CryptoUtils.cs
Comment (PaulHigin) [inline src/System.Management.Automation/utils/CryptoUtils.cs]: Localizable messages need to go into `SecuritySupportStrings.resx`.

**Review by PaulHigin — CHANGES_REQUESTED** [review]
Overall this looks good to me.  I have tested this on Windows platform and SecureString objects are passed correctly between PowerShell versions.  However, key exchange is not currently working on non-Windows platforms, but I don't think it is due to these changes and is just not supported since it was not needed before.  I'll continue to look into it.

**Comment by SteveL-MSFT** [comment]
Removing `Review - Committee`, after discussing with @PaulHigin this seems like a good change to have, just need security sign-off

**Comment by PaulHigin** [comment]
Key exchange was broken because of remaining #ifdefs in code for non-Windows platform.  Just need to change `OutOfProcServerMediator.cs and serverremotesession.cs` files to get non-Windows support.
We can do that in this PR or a separate one.

File: src/System.Management.Automation/utils/CryptoUtils.cs
Comment (silijon) [reply src/System.Management.Automation/utils/CryptoUtils.cs]: Would your recommend just throwing SecuritySupportStrings.CannotEncryptSecureString for all of these or adding additional string resources to be more specific?

Comment (PaulHigin) [reply src/System.Management.Automation/utils/CryptoUtils.cs]: Sorry for the confusion.  I just meant the above as an example of how to use localizble resx strings for error messages.  In this particular case, the ArgumentException is correct but the message string needs to added to the SecuritySupportStrings.resx file.
```
throw new ArgumentException(SecuritySupportStrings.InvalidBlobBoundary);
```

**Comment by PaulHigin** [comment]
@silijon I have created a PR that addresses my CR comments above, and also enables key exchange for non-Windows platforms.  If you can merge my PR I think we can accept this PR.

https://github.com/SkyKick/PowerShell/pull/1

Thanks for your work on this!

**Review by PaulHigin — APPROVED** [review]

**Comment by SteveL-MSFT** [comment]
@PoshChan please retry linux

**Comment by PoshChan** [comment]
@SteveL-MSFT, successfully started retry of `PowerShell-CI-Linux`

File: src/System.Management.Automation/utils/CryptoUtils.cs
Comment (adityapatwardhan) [inline src/System.Management.Automation/utils/CryptoUtils.cs]:
```suggestion
                throw new ArgumentNullException(nameof(blob));
```

File: src/System.Management.Automation/utils/CryptoUtils.cs
Comment (adityapatwardhan) [inline src/System.Management.Automation/utils/CryptoUtils.cs]:
```suggestion
                throw new ArgumentNullException(nameof(blob));
```

File: src/System.Management.Automation/utils/CryptoUtils.cs
Comment (adityapatwardhan) [inline src/System.Management.Automation/utils/CryptoUtils.cs]:
```suggestion
                throw new ArgumentNullException(nameof(rsa));
```

File: src/System.Management.Automation/utils/CryptoUtils.cs
Comment (adityapatwardhan) [inline src/System.Management.Automation/utils/CryptoUtils.cs]:
```suggestion
                throw new ArgumentNullException(nameof(encryptedKey));
```

**Comment by silijon** [comment]
@PoshChan please retry windows

**Comment by PoshChan** [comment]
@silijon, you are not authorized to request a rebuild

**Comment by vexx32** [comment]
@PoshChan please retry windows

**Comment by PoshChan** [comment]
@vexx32, successfully started retry of `PowerShell-CI-Windows`

**Review by mathewcathey — APPROVED** [review]
Hi

**Review by mathewcathey — CHANGES_REQUESTED** [review]
Yes

**Review by mathewcathey — APPROVED** [review]
Yes

**Comment by PaulHigin** [comment]
@TravisEz13  Where are we on this PR?  Is there anything preventing the merge?

**Review by TravisEz13 — APPROVED** [review]
approving as a maintainer

**Comment by iSazonov** [comment]
@silijon Thanks for your contribution!

**Pull request merged**
